### PR TITLE
Add pj_r api to print raw data

### PR DESCRIPTION
**Pull request by radare**
not sure if it will be better to print raw data as an array of numbers like [2,34,32,159,..] and just keep that string function as is, with null terminated strings only.

another option would be to have another api for non-null terminated strings

### CODE REVIEW

**Comment by ret2libc** [comment]
Why? How do you expect to use this? unit tests please?

**Comment by radare** [comment]
like this:

unsigned char *foo = "\x10\x20\x00\x30\x50";
pj_kd(pj, "data", foo, 5);

**Comment by ret2libc** [comment]
And what's the expected output?

Please, either create a unit test later or fill out a proper PR template so that everybody can understand.

**Comment by radare** [comment]
thats what i am asking, what would you expect?

**Comment by ret2libc** [comment]
> thats what i am asking, what would you expect?

I have no idea :D you proposed the change. Do you need it for something or is it "just to have it"? If it's the latter, just drop the PR imho :) If you need it for something, I guess you know what is the output you expect.

**Comment by radare** [comment]
well i need it for fixing that hacky json constructed with printfs and use a generic function that serializes binary data in json

**Comment by radare** [comment]
yes, when fixing some json bugs i found there's a call in zj that was doing this and making the whole thing bad

**Comment by radare** [comment]
@ret2libc review again

File: libr/include/r_util/pj.h [hunk 1]
```diff
@@ -39,6 +39,8 @@ R_API PJ *pj_kd(PJ *j, const char *k, double d);
 R_API PJ *pj_kf(PJ *j, const char *k, float d);
 R_API PJ *pj_kb(PJ *j, const char *k, bool v);
 R_API PJ *pj_null(PJ *j);
+R_API PJ *pj_r(PJ *j, const unsigned char *k, size_t len);
```
Comment (ret2libc) [inline libr/include/r_util/pj.h]: `k` seems like a very misleading name, as it is usually used for `key`. However this is a value, so i suggest you rename it as `v`.

File: libr/include/r_util/pj.h [hunk 2]
```diff
@@ -39,6 +39,8 @@ R_API PJ *pj_kd(PJ *j, const char *k, double d);
...
+R_API PJ *pj_kr(PJ *j, const char *name, const unsigned char *k, size_t len);
```
Comment (ret2libc) [inline libr/include/r_util/pj.h]: same here. `name` is actually the key AFAIU and `k` is actually the value, so it should be `v` and `len` -> `v_len`.

File: libr/include/r_util/pj.h (hunk 1 above)
Comment (ret2libc) [reply libr/include/r_util/pj.h]: and maybe `len` -> `v_len`

File: libr/util/pj.c [hunk 3]
```diff
@@ -194,12 +194,28 @@ R_API PJ *pj_s(PJ *j, const char *k) {
...
+	size_t i;
+	pj_a (j);
+	for (i = 0; i < len; i++) {
+		pj_i (j, k[i]);
+	}
+	pj_end (j);
+	return j;
+}
+
+R_API PJ *pj_kr(PJ *j, const char *name, const unsigned char *k, size_t len) {
```
Comment (ret2libc) [inline libr/util/pj.c]: Add `r_return_val_if_fail (j && k && v, j);`

File: libr/util/pj.c [hunk 4]
```diff
@@ -194,12 +194,28 @@ R_API PJ *pj_s(PJ *j, const char *k) {
...
 		free (ek);
 	} else {
-		eprintf ("damn\n");
+		eprintf ("cannot escape string\n");
 	}
 	pj_raw (j, "\"");
 	return j;
 }
 
+R_API PJ *pj_r(PJ *j, const unsigned char *k, size_t len) {
```
Comment (ret2libc) [inline libr/util/pj.c]: same here, add `r_return_val_if_fail (j && k, j);`

**Comment by radare** [comment]
lets wait for travis and merge

File: libr/include/r_util/pj.h [hunk 5]
```diff
@@ -39,6 +39,8 @@ R_API PJ *pj_kd(PJ *j, const char *k, double d);
 R_API PJ *pj_kf(PJ *j, const char *k, float d);
 R_API PJ *pj_kb(PJ *j, const char *k, bool v);
 R_API PJ *pj_null(PJ *j);
+R_API PJ *pj_r(PJ *j, const unsigned char *v, size_t v_len);
+R_API PJ *pj_kr(PJ *j, const char *name, const unsigned char *v, size_t v_len);
```
Comment (ret2libc) [inline libr/include/r_util/pj.h]: s/name/k

**Pull request merged**